File: src/main/java/com/gpuflight/agent/publisher/KafkaPublisher.java
package com.gpuflight.agent.publisher;

import com.gpuflight.agent.config.JsonSettings;
import com.gpuflight.agent.model.LogWrapper;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import com.gpuflight.agent.config.KafkaConfig;

import java.util.Properties;

public class KafkaPublisher implements Publisher {
    private final KafkaProducer<String, String> producer;

    public KafkaPublisher(KafkaConfig config) {
        var props = new Properties();
        props.put("bootstrap.servers", config.bootstrapServers());
        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        props.put("compression.type", config.compression());
        props.put("linger.ms", String.valueOf(config.lingerMs()));
        this.producer = new KafkaProducer<>(props);
    }

    @Override
    public boolean publish(String topic, String key, LogWrapper log) {
        try {
            String message = JsonSettings.MAPPER.writeValueAsString(log.data());
            // producer.send() is async and only blocks when the internal buffer is full.
            // Blocking on a virtual thread is fine — carrier thread parks, not blocks.
            producer.send(new ProducerRecord<>(topic, key, message)).get();
            return true;
        } catch (Exception e) {
            System.out.println("Kafka publish error: " + e.getMessage());
            return false;
        }
    }

    @Override
    public void close() {
        producer.flush();
        producer.close();
    }
}

File: src/main/kotlin/com/gpuflight/agent/publisher/KafkaPublisher.kt
package com.gpuflight.com.gpuflight.agent.publisher

import com.gpuflight.com.gpuflight.agent.config.JsonSettings.json
import com.gpuflight.com.gpuflight.agent.model.LogWrapper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerRecord
import java.util.Properties

class KafkaPublisher(bootstrapServers: String): Publisher {
    private val producer: KafkaProducer<String, String>

    init {
        val props = Properties().apply {
            put("bootstrap.servers", bootstrapServers)
            put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer")
            put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer")
            put("compression.type", "snappy")
            put("linger.ms", "100")
        }
        producer = KafkaProducer<String, String>(props)
    }

    override suspend fun publish(topic: String, key: String, log: LogWrapper) {
        // Switching to IO dispatcher to avoid blocking the main thread
        // even though send() is async, it can block on buffer full.
        withContext(Dispatchers.IO) {
            val message = json.encodeToString(log.data)
            val record = ProducerRecord(topic, key, message)
            producer.send(record)
        }
    }

    override fun close() {
        producer.flush()
        producer.close()
    }
}